### PR TITLE
3980 primary address demotion

### PR DESCRIPTION
**Pull request by phebal**
Various fixes for bugs in ODE that cause address, phone and email demotions, duplicaiton accross types and other less expected problems.  Bulk of the code is examples that demonstrate the previously untested or poorly tested kinks of the process.

### CODE REVIEW

File: surveys/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+*
+!.gitignore
```
Comment (hannahwhy) [inline surveys/.gitignore:2]: Why are you ignoring the ignore file?  That _should_ be in version control.

File: spec/models/operational_data_extractor/base_spec.rb [hunk 2]
```diff
@@ -268,314 +275,424 @@
...
+      end
 
-          @participant = Factory(:participant)
-          @part_person_link = Factory(:participant_person_link, :participant => @participant, :person => @person)
-          @survey = create_pbs_pregnancy_visit_1_with_birth_institution_operational_data
-          @response_set, @instrument = prepare_instrument(@person, @participant, @survey)
+      it "populates birth address record from instrument responses" do
+        birth_address_and_institution = @pregnancy_visit_extractor.process_birth_institution_and_address(@birth_address_map, @institution_map)
+        birth_address = birth_address_and_institution[0]
+        birth_address.should be_an_instance_of(Address)
```
Comment (hannahwhy) [inline spec/models/operational_data_extractor/base_spec.rb:309]: What requires this to be true?

File: spec/models/operational_data_extractor/base_spec.rb [hunk 3]
```diff
@@ -268,314 +275,424 @@
...
+        birth_address.zip.should == "65445"
+        birth_address.address_type_other.should == "Birth"
+      end
 
-          hospital = NcsCode.for_list_name_and_local_code("ORGANIZATION_TYPE_CL1", 1)
-          state = NcsCode.for_list_name_and_local_code("STATE_CL1", 14)
+      it "returns a created institution" do
+        birth_address_and_institution = @pregnancy_visit_extractor.process_birth_institution_and_address(@birth_address_map, @institution_map)
+        institution = birth_address_and_institution[1]
+        institution.should be_an_instance_of(Institution)
```
Comment (hannahwhy) [inline spec/models/operational_data_extractor/base_spec.rb:321]: Ditto.

For most Ruby operations, the object type is irrelevant.  Some operations do check types (e.g. mutating ActiveRecord associations), but those should be their own expectations:

``` ruby
  it "can be used in Widget#institutions" do
    institution = @pregnancy_visit_extractor.process_ # ...
    widget.institutions << institution

    institutions.length.should == 1
  end
```

File: spec/models/operational_data_extractor/base_spec.rb [hunk 4]
```diff
@@ -268,314 +275,424 @@
...
+        institution.institute_name.should == "FAKE HOSPITAL MEMORIAL"
+      end
 
+      context "birth place name doesn't exist" do
+        it "doesn't return an institution" do
+          @response_set.responses.delete(
+            @response_set.responses.where(
+              :questions => {:data_export_identifier => "#{OperationalDataExtractor::PregnancyVisit::PREGNANCY_VISIT_1_3_INTERVIEW_PREFIX}.BIRTH_PLACE"})
+          )
```
Comment (hannahwhy) [inline spec/models/operational_data_extractor/base_spec.rb:330]: For greater clarity, consider wrapping this pattern in a helper method.

File: spec/models/operational_data_extractor/base_spec.rb [hunk 5]
```diff
@@ -681,12 +813,37 @@ def take_birth_survey(address_one, rs = @response_set)
         ).should == @existing_work_email
       end
 
-      it "creates a new email record if one does not exist " do
-        shared_email_type_code = NcsCode.for_list_name_and_local_code(
-                                                  'EMAIL_TYPE_CL1', 3)
-        new_email = @pbs_eligibility_extractor.get_email(@person,
-                                               'doesnt@exist.net',
-                                               shared_email_type_code)
+      it "creates a new email record can't find one because of address" do
```
Comment (hannahwhy) [inline spec/models/operational_data_extractor/base_spec.rb:816]: This doesn't read well; there seems to be a missing conditional in the sentence.

File: spec/models/operational_data_extractor/base_spec.rb [hunk 6]
```diff
@@ -681,12 +813,37 @@ def take_birth_survey(address_one, rs = @response_set)
...
+          @person,
+          'bad@email.net',
+          @work_email_type_code
+        )
+        new_email.should_not == @existing_work_email
+        new_email.should_not == @existing_email
+        new_email.should be_an_instance_of(Email)
+      end
+
+      it "creates a new email record can't find one because of type in another response_set" do
```
Comment (hannahwhy) [inline spec/models/operational_data_extractor/base_spec.rb:829]: Ditto.

File: spec/models/operational_data_extractor/base_spec.rb [hunk 7]
```diff
@@ -722,14 +881,47 @@ def take_birth_survey(address_one, rs = @response_set)
...
+        @existing_work_phone.phone_rank_code.should == 1
+        @existing_home_phone.phone_rank_code.should == 1
+        @pbs_eligibility_extractor.finalize_telephones(new_work_phone, new_home_phone)
         updated_work_phone = Telephone.find(@existing_work_phone.id)
         updated_work_phone.phone_rank_code.should == 2
+        updated_home_phone = Telephone.find(@existing_home_phone.id)
+        updated_home_phone.phone_rank_code.should == 2
+      end
+
+      it "doesn't demotes existing telephone records in favor of new telephone records of different type" do
```
Comment (hannahwhy) [inline spec/models/operational_data_extractor/base_spec.rb:909]: "demote", not "demotes".

**Comment by phebal** [comment]
Found more problems with ODE that this PR should probably address.  Closing for now to add those fixes.  I'll address the comments before re-opening.

**Pull request closed without merging**